module/os/linux/zfs/zfs_znode_os.c

```diff
@@ -1109,8 +1117,18 @@ zfs_zget(zfsvfs_t *zfsvfs, uint64_t obj_num, znode_t **zpp)
 		if (igrab(ZTOI(zp)) == NULL) {
 			if (zp->z_unlinked)
 				err = SET_ERROR(ENOENT);
-			else
+			else {
 				err = SET_ERROR(EAGAIN);
+				/*
+				 * In writeback path, I_SYNC flag will be set
+				 * and block inode eviction. So we must not
+				 * loop doing igrab in possible writeback
+				 * path, i.e. zfs_get_data, if inode is being
+				 * evicted and I_SYNC is also set.
+				 */
+				if (check_sync && (ZTOI(zp)->i_state & I_SYNC))
```

Comment:
maybe we can move this before the `igrab`, take the `i_lock` and check for `I_FREEING` and `I_SYNC` and if they are both set, bail out of the original `zfs_zget` without the need for `zfs_zget_impl`

Reply:
The only place that needs bail out is if you are in writeback context, either you are the one setting I_SYNC or whoever set it is blocked by you.

If you do bailout for ever caller than you will get error randomly in all sorts of places for no good reason.

Reply:
`I_SYNC` is set in two kernel functions, `writeback_sb_inodes`, `writeback_single_inode`. So if you check for `I_SYNC` and `I_FREEING` and you use `i_lock` correctly as you're supposed to, you get more clean and also reliable solution. How is what am saying wrong _specifically_ and why? Which tons of errors?

Reply:
@snajpa 
I_FREEING and I_SYNC is set on the inode. Every one trying to do zfs_zget will see them. So if you don't have any restriction on caller then every one will bailout and got error as long as eviction and writeback is happening at the same time.

Reply:
well.. I tried playing around with it and it seems like it would be an interesting project to get to the bottom of maybe few other lurking bugs, forcing one to go around and look at how errors from `zfs_zget` are handled all over the codebase - I think there's a thing in direct IO I'll try to tackle if I find time; but that's not relevant to resolving the deadlock, so, yeah, no easy way other than the `_impl` split :)